src/dmtest/gendatablocks.py

```diff
@@ -0,0 +1,663 @@
+        buffer : bytes
+            bytes read from disk
+
+        Returns
+        -------
+        Header
+            Header representing bytes read from disk
+
+        """
+        tag, stream_number, block_number = struct.unpack(HEADER_FORMAT, buffer)
+        return Header(tag.decode('ascii').rstrip('\0'), stream_number, block_number)
+
+class BlockBuffer:
+    """A variable length buffer of data that is written to and read from disk
+
+    It contains a fixed length header and a variable length data portion.
+
+    """
+    def __init__(self, header: 'Header', data: bytes = bytes()):
+        self.header = header
+        self.data = data
+
+    def fill_data(self, compress_size: int, block_size: int):
+        """Fill the variable length part of the buffer
+
+        This function fills the variable length part of the buffer with compressible and
+        random data. The first part is all 1s and is based on how much compression is
+        requested. The remaining data is random data generated using a seed based off of
+        the Header. The code will use similar Headers as a way of generating requested
+        dedupe rates.
+
+        Parameters
+        ----------
+        compress_size : int
+            the requested compression rate
+        block_size : int
+            the block size to write
+
+        """
+        rand_size = block_size - compress_size
+        header_len = Header.len_as_bytes()
+        if compress_size > 0:
+            compress_size = compress_size - header_len
+        else:
+            rand_size = rand_size - header_len
+
+        seed = self.header.get_seed()
+        rng = numpy.random.default_rng(seed)
+        self.data = bytes([255] * compress_size) + rng.bytes(rand_size)
+
+    def make_format(self) -> str:
+        """Generate the appropriate pack/unapack format string
+
+        Returns
+        -------
+        str
+            The pack/unpack string to use
+
+        """
+        return ("@" + str(Header.len_as_bytes()) + "s"
+                + str(len(self.data)) + "s")
+
+    def to_bytes(self) -> int:
+        """get a bytes array representation of the BlockBuffer
+
+        Returns
+        -------
+        bytes
+            the bytes array representation
+
+        """
+        format = self.make_format()
+        return struct.pack(format, self.header.to_bytes(), self.data)
+
+class DataStream:
+    """A generic data stream that operates on a BlockRange"""
+    def __init__(self):
+        self.counter = 0
+
+    def claim(self, buffer: bytes) -> bool:
+        raise NotImplementedError("method claim must be implemented")
+
+    def generate(self, block_number: int, block_size: int) -> bytes:
+        raise NotImplementedError("method generate must be implemented")
+
+    def report(self) -> str:
+        raise NotImplementedError("method report must be implemented")
+
+class BlockStream(DataStream):
+    """A data stream that may contain dedupe and/or compression"""
```

Comment:
"duplicate and/or compressible data" ... dedupe and compression are the actions we will perform on the data